### PR TITLE
Implement planar formulations using new Lagrangian kernels

### LINKED ISSUE #21757: Planar formulations using the new Lagrangian kernels
## Reason

For better compatibility with old kernels, the new kernels should support 
1. plane strain
2. weak plane stress
3. generalized plane strain

## Design

1. Plane strain is easy -- this is the default if we do nothing.
2. For weak plane stress: Couple out-of-plane strain to the strain calculator, add a weak plane stress kernel, and also add off-diag contributions in the stress divergence kernels.
3. We could rely on the existing homogenization system to do generalized plane strain. To avoid code duplication, I propose to add a third option to the constraint type. Currently, "strain" and "stress" are available constraint types for homogenization. If we add a third option "none", we will be able to only include the out-of-plane strain/stress as the constraint. As a bonus, we will be able to do the so-called "generalized plane stress" in addition to generalized plane strain.

## Impact

Better compatibility and interchangeability between the new and old mechanics kernels.

### PR DESCRIPTION
**Pull request by hugary1995**
close #21757

### CODE REVIEW

**Comment by dschwen** [comment]
```
error: unused variable 'ctarget' [-Werror=unused-variable]
         auto && [ctype, ctarget] = constraint;`
```

just do `auto && ctype = constraint.first;` or add `libmesh_ignore(ctarget);`

**Comment by hugary1995** [comment]
Yeah, thanks. I was planning to come back to this once the axisymmetry PR is merged.

File: modules/tensor_mechanics/doc/content/source/materials/lagrangian/ComputeLagrangianWPSStrain.md [hunk 1]
```diff
@@ -0,0 +1,13 @@
+# ComputeLagrangianWPSStrain
+
+!syntax description /Materials/ComputeLagrangianWPSStrain
+
+## Overview
+
+The `ComputeLagrangianWPSStrain` inherits from [`ComputeLagrangianStrain`](ComputeLagrangianStrain.md). Only two displacement variables are coupled to compute the in-plane strain components. The out-of-plane strain is provided by another nonlinear variable served as a Lagrange multiplier to weakly enforce the plane stres condition.
```
Comment (dschwen) [inline modules/tensor_mechanics/doc/content/source/materials/lagrangian/ComputeLagrangianWPSStrain.md:7]:
```suggestion
The `ComputeLagrangianWPSStrain` inherits from [`ComputeLagrangianStrain`](ComputeLagrangianStrain.md). Only two displacement variables are coupled to compute the in-plane strain components. The out-of-plane strain is provided by another nonlinear variable served as a Lagrange multiplier to weakly enforce the plane stress condition.
```

File: modules/tensor_mechanics/include/kernels/lagrangian/LagrangianStressDivergenceBase.h [hunk 2]
```diff
@@ -49,12 +49,15 @@ class LagrangianStressDivergenceBase
   virtual Real computeQpJacobian() override;
   virtual Real computeQpOffDiagJacobian(unsigned int jvar) override;
 
-  /// Derivatives of the residual w.r.t. the displacement dofs
+  // Derivatives of the residual w.r.t. the displacement dofs
```
Comment (dschwen) [inline modules/tensor_mechanics/include/kernels/lagrangian/LagrangianStressDivergenceBase.h:52]: why not doxygen?

Comment (hugary1995) [reply modules/tensor_mechanics/include/kernels/lagrangian/LagrangianStressDivergenceBase.h:52]: I didn't delete the doxygen, did I?

**Comment by hugary1995** [comment]
@dschwen anything else you want me to address?

**Review by dschwen — APPROVED** [review]

**Pull request merged**